### PR TITLE
[SPARK-32131][SQL] Fix AnalysisException messages at UNION/EXCEPT/MINUS operations

### PR DESCRIPTION
**Pull request by GuoPhilipse**
### What changes were proposed in this pull request?
fix error exception messages during exceptions on Union and set operations

### Why are the changes needed?
Union and set operations can only be performed on tables with the compatible column types,while when we have more than two column, the exception messages will have wrong column index.

Steps to reproduce:

```
drop table if exists test1; 
drop table if exists test2; 
drop table if exists test3;
create table if not exists test1(id int, age int, name timestamp);
create table if not exists test2(id int, age timestamp, name timestamp);
create table if not exists test3(id int, age int, name int);
insert into test1 select 1,2,'2020-01-01 01:01:01';
insert into test2 select 1,'2020-01-01 01:01:01','2020-01-01 01:01:01'; 
insert into test3 select 1,3,4;
```

Query1:
```sql
select * from test1 except select * from test2;
```
Result1:
```
Error: org.apache.spark.sql.AnalysisException: Except can only be performed on tables with the compatible column types. timestamp <> int at the second column of the second table;; 'Except false :- Project [id#620, age#621, name#622] : +- SubqueryAlias `default`.`test1` : +- HiveTableRelation `default`.`test1`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [id#620, age#621, name#622] +- Project [id#623, age#624, name#625] +- SubqueryAlias `default`.`test2` +- HiveTableRelation `default`.`test2`, org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe, [id#623, age#624, name#625] (state=,code=0)
```

Query2:

```sql
select * from test1 except select * from test3;
```

Result2:

```
Error: org.apache.spark.sql.AnalysisException: Except can only be performed on tables with the compatible column types
 int <> timestamp at the 2th column of the second table;
```

the above query1 has the right exception message
the above query2 have the wrong errors information, it may need to change to the following

```
Error: org.apache.spark.sql.AnalysisException: Except can only be performed on tables with the compatible column types.
int <> timestamp at the  third column of the second table
```

### Does this PR introduce _any_ user-facing change?
NO

### How was this patch tested?
unit test

### CODE REVIEW

**Comment by holdenk** [comment]
Good catch. LGTM but I'll leave it for a bit of a SQL committer has any thoughts.

**Comment by HyukjinKwon** [comment]
ok to test

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 1]
```diff
@@ -831,4 +831,77 @@ class AnalysisSuite extends AnalysisTest with Matchers {
...
+      AttributeReference("c", TimestampType)(),
+      AttributeReference("d", FloatType)())
+
+    val fourthTable = LocalRelation(
+      AttributeReference("a", StringType)(),
+      AttributeReference("b", DoubleType)(),
+      AttributeReference("c", IntegerType)(),
+      AttributeReference("d", TimestampType)())
+
+    val a1 = firstTable.output(0)
```
Comment (dilipbiswal) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala:861]: @GuoPhilipse Variables a1, b1, c1, d1 not used ? Were you planning to use it in the test later ?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 2]
```diff
@@ -831,4 +831,77 @@ class AnalysisSuite extends AnalysisTest with Matchers {
...
+    val b1 = firstTable.output(1)
+    val c1 = firstTable.output(2)
+    val d1 = firstTable.output(3)
+
+    val a2 = secondTable.output(0)
```
Comment (dilipbiswal) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala:866]: ditto

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 3]
```diff
@@ -831,4 +831,77 @@ class AnalysisSuite extends AnalysisTest with Matchers {
...
+    val b2 = secondTable.output(1)
+    val c2 = secondTable.output(2)
+    val d2 = secondTable.output(3)
+
+    val a3 = thirdTable.output(0)
```
Comment (dilipbiswal) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala:871]: ditto

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 4]
```diff
@@ -831,4 +831,77 @@ class AnalysisSuite extends AnalysisTest with Matchers {
...
+    val b3 = thirdTable.output(1)
+    val c3 = thirdTable.output(2)
+    val d3 = thirdTable.output(3)
+
+    val a4 = fourthTable.output(0)
```
Comment (dilipbiswal) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala:876]: ditto

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala (hunk 1 above)
Comment (GuoPhilipse) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala:861]: em,,,let me remove it

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 5]
```diff
@@ -337,7 +337,8 @@ trait CheckAnalysis extends PredicateHelper {
             def ordinalNumber(i: Int): String = i match {
               case 0 => "first"
               case 1 => "second"
-              case i => s"${i}th"
+              case 2 => "third"
+              case i => s"${i + 1}th"
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:341]: hehehe, nice catch ;)

**Comment by maropu** [comment]
nit: `test2` not used in the PR description?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 6]
```diff
@@ -831,4 +831,57 @@ class AnalysisSuite extends AnalysisTest with Matchers {
       }
     }
   }
+
+  test("SPARK-32131 Fix wrong column index when we have more than two columns" +
```
Comment (maropu) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala:835]: nit: `SPARK-32131` -> `SPARK-32131:`

**Review by maropu — APPROVED** [review]

**Comment by GuoPhilipse** [comment]
> nit: `test2` not used in the PR description?

have updated in the PR, it is a normal case ,it is used for comparing between the normal and abnormal cases.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala (hunk 6 above)
Comment (GuoPhilipse) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala]: Thanks maropu ,btw, I did not see the test build start to build, do you know why?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala (hunk 4 above)
Comment (GuoPhilipse) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala]: Thanks @dilipbiswal for your review.

**Comment by dilipbiswal** [comment]
LGTM

**Comment by GuoPhilipse** [comment]
retest this please.

**Comment by GuoPhilipse** [comment]
It seesm i cannot trigger the test build, @dilipbiswal @maropu @HyukjinKwon  @holdenk ,do you have any ideas?

**Comment by holdenk** [comment]
Jenkins ok to test

**Comment by holdenk** [comment]
Jenkins retest this please

**Comment by maropu** [comment]
ok to test

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by dongjoon-hyun** [comment]
Thank you all! This lands at `master/3.0/2.4`.

**Pull request closed without merging**